ultravox/tools/ds_tool.py

```diff
@@ -0,0 +1,172 @@
+class TextGenerationTask:
+    language_model: str = simple_parsing.field(default="gpt-4o", alias="-m")
+    new_column_name: str = simple_parsing.field(default="explanation", alias="-c")
+    template: str = DEFAULT_TEXTGEN_TEMPLATE
+
+
+# This script is used to either generate audio samples from text using a TTS model, or to generate text samples using a text generation model.
+# Ex: just ds_tool -T tts -d google/boolq -u fixie-ai/boolq-audio -c question -a audio
+# Ex: just ds_tool -T textgen -d fixie-ai/boolq-audio -u fixie-ai/boolq-audio -c explanation
+# Ex: just ds_tool -T textgen -d ylacombe/expresso -u fixie-ai/expresso -c continuation \
+#         --template "\"Continue the following sentence in a way that reflects a ‘{style}’ tone in a coherent style:\n{text}\""
+@dataclasses.dataclass
+class DatasetToolArgs:
+    dataset_name: str = simple_parsing.field(alias="-d")
+    dataset_subset: str = simple_parsing.field(default="default", alias="-S")
+    dataset_split: Optional[str] = simple_parsing.field(default=None, alias="-s")
+
+    num_samples: Optional[int] = simple_parsing.field(default=None, alias="-n")
+    num_workers: int = simple_parsing.field(default=16, alias="-w")
+
+    upload_name: Optional[str] = simple_parsing.field(default=None, alias="-u")
+    upload_branch: Optional[str] = simple_parsing.field(default="main", alias="-b")
+    num_shards: Optional[int] = simple_parsing.field(default=None, alias="-N")
+
+    token: Optional[str] = simple_parsing.field(default=None, alias="-t")
+
+    task: Union[TtsTask, TextGenerationTask] = simple_parsing.subgroups(
+        {"tts": TtsTask, "textgen": TextGenerationTask},  # type: ignore
+        default_factory=TtsTask,
+        alias="-T",
+    )
+
+
+def _tts_sample(
+    sample, col_name: str, audio_col_name: str, voice: str, sample_rate: int
+):
+    text = sample[col_name]
+    text = text["text"] if isinstance(text, dict) else text
+    sample[audio_col_name] = TTS_CLIENT.tts(text, voice=voice, sample_rate=sample_rate)
+    return sample
+
+
+def _tts_split(
+    ds_split: datasets.Dataset, task: TtsTask, num_proc: Optional[int] = None
+):
+    col_name = task.column_name
+    audio_col_name = task.audio_column_name or f"{col_name}_audio"
+    print(f'TTS mapping "{col_name}" to "{audio_col_name}"...')
+
+    return ds_split.map(
+        _tts_sample,
+        num_proc=num_proc,
+        fn_kwargs={
+            "col_name": col_name,
+            "audio_col_name": audio_col_name,
+            "voice": task.voice,
+            "sample_rate": task.sample_rate,
+        },
+    ).cast_column(audio_col_name, datasets.Audio(sampling_rate=task.sample_rate))
+
+
+def _text_gen_sample(sample, template: str, col_name: str, language_model: str):
+    input_text = template.format(**sample)
+    response = CHAT_CLIENT.chat.completions.create(
+        model=language_model,
+        messages=[{"role": "user", "content": input_text}],
+        max_tokens=128,
```

Comment:
probably going to want to customize these in the future, I'd make them properties of the task